.scripts/ls.py

```diff
@@ -263,37 +285,106 @@ def _snapshot_repo(ref: str | None):
+    return _get_interface_version(path, root=root)
+
+
+def _get_summary(category: str, root: pathlib.Path, path: pathlib.Path) -> str:
+    if category == 'packages':
+        return _get_description(category, root, path)
+    assert category == 'interfaces'
+    return _interface_yaml(path, root=root).get('summary', '').strip()
+
+
+def _get_description(category: str, root: pathlib.Path, path: pathlib.Path) -> str:
+    if category == 'packages':
+        return _pyproject_toml(path, root=root)['project']['description'].strip()
+    assert category == 'interfaces'
+    return _interface_yaml(path, root=root).get('description', '').strip()
+
+
+def _get_lib_name(category: str, root: pathlib.Path, path: pathlib.Path) -> str:
+    if category == 'packages':
+        if path.name == '.package':
+            # .package -> () -> 'charmlibs'
+            # interfaces/.package -> ('interfaces') -> 'charmlibs.interface'
+            parts, _ = path.parts
+        else:
+            # For special cases like '.tutorial' -> ('tutorial') -> 'charmlibs.tutorial'
+            parts = tuple(p.removeprefix('.') for p in path.parts)
+        return '.'.join(('charmlibs', *parts)).replace('-', '_')
+    assert category == 'interfaces'
+    return _interface_yaml(path, root=root).get('lib', '').strip()
+
+
+def _get_docs_url(category: str, root: pathlib.Path, path: pathlib.Path) -> str:
     if category == 'packages':
-        return _get_dist_name(root / path)
+        url = _pyproject_toml(path, root=root)['project']['urls'].get('Documentation', '')
+        return url.strip()
     assert category == 'interfaces'
-    return (root / path).name
+    return f'https://documentation.ubuntu.com/charmlibs/reference/interfaces/{path.name}/'
 
 
-def _get_dist_name(package: pathlib.Path) -> str:
+def _get_dist_name(package: pathlib.Path, root: pathlib.Path = _REPO_ROOT) -> str:
     """Load distribution package name from pyproject.toml and normalize it."""
-    with (package / 'pyproject.toml').open('rb') as f:
-        return _normalize(tomllib.load(f)['project']['name'])
+    name = _pyproject_toml(package, root=root)['project']['name']
+    return _normalize(name.strip())
+
+
+@functools.cache
+def _get_package_version(package: pathlib.Path, root: pathlib.Path = _REPO_ROOT) -> str:
+    name = _get_dist_name(package, root=root)
+    script = f'import importlib.metadata; print(importlib.metadata.version("{name}"))'
+    cmd = ['uv', 'run', '--no-project', '--with', root / package, 'python', '-c', script]
+    return subprocess.check_output(cmd, cwd=root, text=True).strip()
+
+
+@functools.cache
+def _get_interface_version(path: pathlib.Path, root: pathlib.Path = _REPO_ROOT) -> str:
+    versions = [v.name for v in (root / path / 'interface').glob('v[0-9]*')]
+    return max((v.removeprefix('v') for v in versions), key=int)
+
+
+@functools.cache
+def _lib_urls() -> dict[str, str]:
+    result: dict[str, str] = {}
+    for p in 'general-libs.csv', 'interface-libs.csv':
```

Comment:
I wonder if we should be loading content from `.docs` here. It feels like `.docs` should be self-contained (or write-only, I guess).

If we need to get data from this CSV (which is starting to feel like it ought to be a sqlite3 db 😆) then it feels like it ought to be placed elsewhere and the docs could either have a symlink or access it from elsewhere.

Reply:
Yeah, not sure where would be best. Maybe the root does make sense, because they are basically summarising everything contained here (plus more)?